src/broadcast/broadcast.rs

```diff
@@ -284,21 +380,130 @@ impl<N: NodeIdT> Broadcast<N> {
             // Enqueue a broadcast of a Ready message.
             step.extend(self.send_ready(hash)?);
         }
+        // Upon receiving 2f + 1 matching Ready(h) messages, send full
+        // `Echo` message to every node who hasn't sent us a `CanDecode`
+        if self.count_readys(hash) == 2 * self.netinfo.num_faulty() + 1 {
+            step.extend(self.send_echo_remaining(hash)?);
+        }
+
         Ok(step.join(self.compute_output(hash)?))
     }
 
-    /// Sends an `Echo` message and handles it. Does nothing if we are only an observer.
-    fn send_echo(&mut self, p: Proof<Vec<u8>>) -> Result<Step<N>> {
-        self.echo_sent = true;
+    /// Sends `Echo` message to all left nodes and handles it.
+    fn send_echo_left(&mut self, p: Proof<Vec<u8>>) -> Result<Step<N>> {
         if !self.netinfo.is_validator() {
             return Ok(Step::default());
         }
         let echo_msg = Message::Echo(p.clone());
-        let step: Step<_> = Target::All.message(echo_msg).into();
+        let mut step = Step::default();
+        // `N - 2f + g` node ids to the left of our_id (excluding our_id and proposer_id)
+        // after arranging all node ids in a circular list.
+        let left = self
+            .netinfo
+            .all_ids()
+            .cycle()
+            .skip_while(|x| *x != self.our_id())
+            .take(self.netinfo.num_correct() - self.netinfo.num_faulty() + self.fault_estimate)
+            .skip(1);
+        for id in left {
+            let msg = Target::Node(id.clone()).message(echo_msg.clone());
+            step.messages.push(msg);
+        }
         let our_id = &self.our_id().clone();
         Ok(step.join(self.handle_echo(our_id, p)?))
     }
 
+    /// Sends `Echo` message to remaining nodes who haven't sent `CanDecode`
+    fn send_echo_remaining(&mut self, hash: &Digest) -> Result<Step<N>> {
+        self.echo_sent = true;
+        if !self.netinfo.is_validator() {
+            return Ok(Step::default());
+        }
+
+        // Simply return as we can decode the message anyway as we have N - 2f shards.
+        if !self.echos.contains_key(self.our_id()) {
+            return Ok(Step::default());
+        }
+
+        // Can't send `Echo` if we haven't received a `Value` message for that hash from Proposer.
+        if let Some((h, _)) = self.echos.get(self.our_id()) {
+            if h != hash {
+                return Ok(Step::default());
+            }
+        }
+
+        if let Some((_, None)) = self.echos.get(self.our_id()) {
+            return Ok(Step::default());
+        }
+        let (_, p) = &self.echos[self.our_id()];
+        let p = p.clone().unwrap();
+        let echo_msg = Message::Echo(p.clone());
```

Comment:
No need to clone `p` again: We only use its hash below, which is already in the `hash` variable.